More logic changes to error reporting, cleanup

agjohnson opened the pull request:
* Alter project.task logic and include multiple points of error reporting during
  build task
* Build task returns true or false, not project data, which is unused anyways
* New pattern for error messages
* Localize error messages?
* Make more specific error messages for repository error and project error
* Remove unnecessary git/hg/bzr/svn exit codes and superfluous command data from
  user errors
* Add handling of git prompt for git 2.3+
* Add errors for docker connection
* Handle reporting on not able to connect to docker

agjohnson commented:
Marked WIP, because:

- [ ] Regression in reporting build finished in `run_setup()`
- [ ] Not unittested yet
- [ ] Need to play with exception translation pattern -- ultimately not worried about this yet, it will be an eventual improvement however.
- [ ] Need to do more to catch an API exception with docker on memory issues and report this
- [ ] Should actually write unittests around all of this

agjohnson commented:
Also, I'm thinking that we need to move the vcs execution into a build environment, the error reporting with git stderr is sloppy, but there's no reason we can't just show this step to users. This will be a separate ticket.

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -550,15 +553,17 @@ def __exit__(self, exc_type, exc_value, tb):
+        # Catch direct failures from Docker API, but also requests exceptions
+        # with the HTTP request
+        except (DockerAPIError, ConnectionError):
             log.error(LOG_TEMPLATE
```
isn't it better here to use `log.exception` ?

humitos commented on readthedocs/projects/exceptions.py:
```diff
@@ -1,15 +1,46 @@
-    """Failure to import a project from a repository."""
 
-    pass
+class ProjectConfigurationError(BuildEnvironmentError):
```
I think this class is missing the `get_default_message` method.

agjohnson replied:
It should be inheriting this from `BuildEnvironmentException`

humitos commented on readthedocs/vcs_support/backends/git.py:
```diff
@@ -58,10 +59,8 @@ def repo_exists(self):
     def fetch(self):
         code, _, err = self.run('git', 'fetch', '--tags', '--prune')
```
`err` could be `_` now that's not used

humitos commented on readthedocs/vcs_support/backends/git.py:
```diff
@@ -79,16 +78,7 @@ def clone(self):
         code, _, err = self.run('git', 'clone', '--recursive', '--quiet',
```
`err` could be `_` now that's not used

humitos commented on readthedocs/vcs_support/backends/git.py:
```diff
@@ -79,16 +78,7 @@ def clone(self):
-                    sterr=err
-                )
-            )
+            raise RepositoryError
```
I think this is too much simpler that what we had, but what happen if there is another problem different than a private repository. I mean, the error message is very generic and we are not checking a specific error code:

For example:

```
[humitos@julia:tmp]$ GIT_TERMINAL_PROMPT=0 git clone https://github.com/readthedocs/readthedocs-corporate
Clonando en 'readthedocs-corporate'...
fatal: could not read Username for 'https://github.com': terminal prompts disabled
[humitos@julia:tmp]$  echo $?
128
```

Ha, it doesn't seem to have a different error code:
```
[humitos@julia:tmp]$ LANG=C git clone git@github.com:rtfd/readthedocs
Cloning into 'readthedocs'...
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
[humitos@julia:tmp]$ echo $?
128
```

Anyway, I think this is not really important if we don't have history/logs failing at these points because something different than a private repository.

agjohnson replied:
I agree, but this would be out of scope for this work, and would be best served by surfacing the commands we execute to users. Currently, the checkout step is hidden because it uses it's own shell execution. This could be easily moved into a BuildCommand executed through the setup environment. This way, users can actually see the output in a meaningful way.

humitos replied:
I like your proposal of moving it inside a BuildCommand and I agree it could be done in another PR.

humitos commented on readthedocs/projects/tasks.py:
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
             setup_successful = self.run_setup(record=record)
-            if setup_successful:
-                self.run_build(record=record, docker=docker)
+            if not setup_successful:
```
I _think_ there is something missing here. If the build fails because of the setup at this point: https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R198

we are not raising any exception, so `run_setup` method just returns `False` and the task return `False` but nobody is updating the `self.setup_env.build['error']` as we do when unhandled exception is catched at https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R137 or https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R154

Also, I think that's better to move this line https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R207 inside this if, so we try to update the build state in not so many places

agjohnson replied:
Links above have disconnected here, so hard to follow what's happening, but i think i understand this feedback.

So update of the build now always happens on build env `__exit__`, so we can just return false from the task.

In the case that another exception is raised outside the env, we still run `update_build` from the exception.

ericholscher replied:
Is this logic ^^ documented somewhere?

agjohnson replied:
Yup, on the environment class. I can expand more.

ericholscher commented:
Feel like this is a biggish change, which might conflict over time -- should try and get tests passing and work done, and then I'll review it.

agjohnson commented:
Test failure was unrelated.

agjohnson commented:
Unittests around doc build environments have been made working, and more have been added.

a9c2a22 is lint clean up, review that separately.

ericholscher commented:
How do I review it seperately?

humitos commented:
You can select a range of commits to review from the top left dropdown: https://github.com/rtfd/readthedocs.org/pull/3310/files/0fd836669a2a3f87866634c534d3a32c7dff462b

(just the 2 commits to review with logic)

humitos commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -121,16 +121,22 @@ def append_conf(self, **__):
+        except (ProjectConfigurationError, IOError):
             trace = sys.exc_info()[2]
-            six.reraise(ProjectImportError('Conf file not found'), None, trace)
+            six.reraise(
```
Why do we need this `six.reraise`? It's not the same than just `raise ProjectImportError(...)`?

agjohnson replied:
I'm not sure either, the docs aren't clear to me:
https://pythonhosted.org/six/#six.reraise

I'll leave it, assuming someone with more knowledge of py2/3 compat did this :)

ericholscher commented:
@humitos thanks!

ericholscher commented on readthedocs/doc_builder/environments.py:
```diff
@@ -294,7 +297,7 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, tb):
         ret = self.handle_exception(exc_type, exc_value, tb)
-        self.build['state'] = BUILD_STATE_FINISHED
+        self.update_build(BUILD_STATE_FINISHED)
```
Won't this post that status to the API? Seems like we don't want to publish here, but only below if `finalize` is `False`.

agjohnson replied:
The `finalize` check is in `update_build`, so it should only post if `finalize` is `True`

ericholscher replied:
Gah ok. Lost that in the review.

ericholscher commented on readthedocs/doc_builder/environments.py:
```diff
@@ -521,8 +525,15 @@ def __enter__(self):
+        except BuildEnvironmentError:
+            # There may have been a problem connecting to Docker altogether, or
+            # some other handled exception here.
+            self.__exit__(*sys.exc_info())
```
Will this not automatically call `__exit__`?

agjohnson replied:
It won't, because this is raised in `__enter__`. Errors raised in `__enter__` raise to the parent context, so we have to explicitly clean up here.

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -275,15 +276,17 @@ class BuildEnvironment(object):
     :param build: Build instance
     :param record: Record status of build object
     :param environment: shell environment variables
+    :param finalize: finalize the build by setting a finished state on exit
```
this sentence is `True` _only_ if the `update_build` method is called with the `BUILD_STATE_FINISHED` status, otherwise that method will just update the build, but not finalize it

agjohnson replied:
Ah true. I didn't like `finalize` either. I'll rename this `commit` to match the Django usage. Would this change make more sense?

humitos replied:
maybe the `commit` should be in the `update_build` method to avoid the other problems that I mentioned

ericholscher commented on readthedocs/doc_builder/environments.py:
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
-        self.container = None
-        self.build['state'] = BUILD_STATE_FINISHED
+        ret = self.handle_exception(exc_type, exc_value, tb)
+        self.update_build(BUILD_STATE_FINISHED)
```
Feels like we're posting this to the API a lot more with these changes -- should we not respect `finalize` everywhere we set build state to finished?

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
+            try:
+                client.kill(self.container_id)
+            except DockerAPIError:
+                pass
```
what about logging something here (DEBUG level, maybe)?

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
+                log.info('Removing container %s', self.container_id)
+                client.remove_container(self.container_id)
+            # Catch direct failures from Docker API, but also requests exceptions
+            # with the HTTP request. These should not
```
"These should not" ?

ericholscher commented on readthedocs/projects/exceptions.py:
```diff
@@ -1,15 +1,46 @@
-    """Failure to import a project from a repository."""
 
-    pass
+class ProjectConfigurationError(BuildEnvironmentError):
```
It seems we have a `BuildEnvironmentException` and a `BuildEnvironmentError`, what is the difference? Seems confusing..

agjohnson replied:
Exception is the base class, Error is the specific class that kills a build and reports to a user, Warning also is absed on Exception and doesn't kill the build.

ericholscher commented on readthedocs/projects/exceptions.py:
```diff
@@ -1,15 +1,46 @@
+    PRIVATE_REPO = _(
+        'There was a problem connecting to your repository, '
+        'ensure that your repository URL is correct.'
+    )
```
If it's a private repo, shouldn't we warn them here? Why are these messages the same?

agjohnson replied:
We aren't thoroughly testing if this was a private repo or not, we'd have to grep the git output. @humitos raised a similar question and the best answer is to execute checkout in a build environment and report the commands, not parse them to determine what the error was.

Messages aren't the same, but usage could be clearer. I should probably rename these constants, as `PRIVATE_REPO` is raised if we support private repositories -- that is, we tell the user "check your url, because it wasn't the repository privacy that caused the problem. `PUBLIC_REPO` error is "check your url or your project privacy, because we don't support private repositories."

In fact, I'll update the copy here to be more explicit about repo privacy.

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
+                    ),
+                )
+            self.container = None
+        except BuildEnvironmentError:
```
I'm not following this. I think it's not possible to `BuildEnvironmentError` be raised in the `try:` block here. What could be the case?

agjohnson replied:
`get_client` raises a `BuildEnvironmentError` on failure here. I'll rethink refactoring this as well though, perhaps a more specific exception, or allowing the docker exception to bubble up makes more sense.

ericholscher commented on readthedocs/projects/tasks.py:
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
+                    'Please include the build id ({build_id}) in any bug reports.'.format(
+                        build_id=build_pk
+                    ))
+                self.build_env.update_build(BUILD_STATE_FINISHED)
```
Another place we're updating build finished..?

agjohnson replied:
It seems this is what we want, no? This is the fallback that was added in case there are any exceptions outside of the build environment `__enter__`, normal context, or in `__exit__`, such as in `run_build`.

ericholscher replied:
Ah, ok. I got lost when reviewing this for some reason, and didn't realize where everything was getting called from.

ericholscher commented on readthedocs/rtd_tests/tests/test_doc_building.py:
```diff
@@ -266,7 +591,23 @@ def test_container_timeout(self):
+            'setup': u'',
+            'output': u'',
+            'state': u'finished',
+            'builder': u'foo'
```
All these listings of massive argument lists seem brittle (it will break whenever we add/change/remove/ an argument). Can we just check for the actual value we care about being called, presumably `error`?

humitos replied:
Test are failing because of this, in fact:

    'length': <ANY>

Besdies, it's complicated to find the differences when it fails. In case you want to leave as it is now, I'd suggest to use `assertDictEqual`

ericholscher commented on readthedocs/vcs_support/backends/git.py:
```diff
@@ -56,12 +56,9 @@ def repo_exists(self):
-                "Failed to get code from '%s' (git fetch): %s\n\nStderr:\n\n%s\n\n" % (
-                    self.repo_url, code, err)
-            )
+            raise RepositoryError
```
Do we want to lose the specifics of these errors?

agjohnson replied:
I think surfacing an error that is helpful provides more user guidance -- that is, showing a failed checkout command isn't clear to the user that the problem is their configuration. I think the ultimate fix here is to do both and show the command execution as a `BuildCommand`. I removed this for now, for consistency on errors and because displaying the stderr output like this looked sloppy. I think we can pretty quickly move towards combining execution environments though.

humitos replied:
I think that having both in the future is better.

I like the idea of communicate the "common problem" to the user in a very simplified way, but also show the output (that could be used when debugging something and could help the user to understand it better -also if the problem is something different that we are reporting as simplified way)

humitos commented on readthedocs/projects/tasks.py:
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
-            if unhandled_failure:
-                self.build_env.build['error'] = unhandled_failure
-            self.build_env.update_build(BUILD_STATE_FINISHED)
+            self.setup_env.update_build(BUILD_STATE_FINISHED)
```
Since `self.setup_env` was instantiated with `finalize=False` this won't work.

agjohnson replied:
Ah yes, good catch!

humitos commented on readthedocs/projects/tasks.py:
```diff
@@ -297,19 +313,10 @@ def setup_vcs(self):
         self.setup_env.update_build(state=BUILD_STATE_CLONING)
```
Also a problem here. The build won't update with this status since it's `finalize=False`.

agjohnson commented:
Okay! There is even more logic now. I believe I've addressed all of the feedback and have dropped in a bunch more docs while this is fresh. One last pass would be a good idea

humitos commented:
I'd like to take a look a it tomorrow

El 30 nov. 2017 7:20 p. m., "Anthony" <notifications@github.com> escribió:

Okay! There is even more logic now. I believe I've addressed all of the
feedback and have dropped in a bunch more docs while this is fresh. One
last pass would be a good idea

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/rtfd/readthedocs.org/pull/3310#issuecomment-348363713>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAO7sPgtnD0HvbdUwdy6V-2uf1OnhTI1ks5s70ZDgaJpZM4QpJDD>
.

agjohnson commented:
Pinging @humitos ☎️

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -267,23 +268,39 @@ class BuildEnvironment(object):
+
+    We only update the build through the API in one of three cases:
+
+    * The build is not done and needs incremental builds
```
What is an incremental build?

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -267,23 +268,39 @@ class BuildEnvironment(object):
     :param build: Build instance
     :param record: Record status of build object
     :param environment: shell environment variables
+    :param commit: update the build object via API if the build was successful
```
I think this is still confusing.

When `commit=True` (the default) and we are in the middle of the build, the build won't be updated. So, I think this should have a better name (if that is possible) like `last_command`, since I think it's more representative of what it measn.

Anyway, I don't see the best solution very clear.

agjohnson replied:
probably the most explicit we can is `update_on_success`

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -386,9 +403,12 @@ def done(self):
-        for the build)
+        This step is skipped if we aren't recording the build. To avoid
+        recording successful builds yet (for instance, running setup commands for
+        the build), set the ``commit`` argument on environment instantiation.
```
set the commit argument to `False`

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -417,37 +437,47 @@ def update_build(self, state=None):
+            try:
+                api_v2.build(self.build['id']).put(self.build)
+            except HttpClientError as e:
+                log.error("Unable to post a new build: %s", e.content)
```
`Unable to update the build:` and maybe add the build `id` also to the log

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -537,28 +574,43 @@ def __enter__(self):
+                client.kill(self.container_id)
+            except DockerAPIError:
+                log.exception(
+                    'Unable to remove container: id=%s',
```
`Unable to kill the container`

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -576,13 +628,21 @@ def get_client(self):
+                LOG_TEMPLATE.format(
+                    project=self.project.slug,
+                    version=self.version.slug,
+                    msg='Could not connection to Docker API',
```
Could not _connect_

humitos commented on readthedocs/doc_builder/environments.py:
```diff
@@ -655,11 +715,22 @@ def create_container(self):
+                    msg=e,
+                ),
+            )
+            raise BuildEnvironmentError('There was a problem connecting to Docker')
```
Here, the problem is similar than https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-ca52b098301dd315a834b3556ab9a7d5R638 but in this case we are communicating the Docker error to the user.

agjohnson commented:
@humitos Thanks for the thorough feedback! I'm just waiting on a final CI test here. If changes look good, lets merge!
